### PR TITLE
Tile and distribute linalg.generic in DispatchLinalgOnTensor

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp [hunk 1]
```diff
@@ -312,6 +312,45 @@ static LogicalResult getConfigForCooperativeMatmul(
...
+                                LaunchConfigInfo &config) {
+  int64_t subgroupSize =
+      targetEnv.getResourceLimits().subgroup_size().getValue().getSExtValue();
+  config.workgroupSize[0] = subgroupSize;
+  config.workgroupSize[1] = 1;
+  config.workgroupSize[2] = 1;
+  ShapedType outputShape = op.getOutputShapedType(0);
+
+  SmallVector<int64_t, 4> sizes;
+  // When Vectororization is not enabled the convertToGPU pass assumes that that
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:330]: Misleading comment. the `ConvertToGPU` pass does not assume a tile size right? I am confused cause the launch config isnt even called in the `ConvertToGPU` Pass.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp [hunk 2]
```diff
@@ -312,6 +312,45 @@ static LogicalResult getConfigForCooperativeMatmul(
...
+  // tile size is equal to the workgroup size.
+  if (options.enableVectorization) {
+    sizes.append({4 * subgroupSize, 2 * subgroupSize});
+  }
+  sizes.push_back(subgroupSize);
+  int64_t lowerTs = config.workgroupSize[0];
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:336]: Could you add comments to whats happening here. Hard to parse this a bit.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 3]
```diff
@@ -612,13 +612,13 @@ struct TileAndDistributeOnTensorsPattern
...
         llvm::map_range(linalgOp.createLoopRanges(rewriter, loc),
                         [](Range r) { return r.size; }));
-    // NOTE: Special treatment for convolution, which have more than 3 parallel
-    // dimensions. We want to ignore the batch dimension and tile along the
-    // next three.
-    // TODO(#5048): figure out a better way to avoid this special case.
-    if (isa<linalg::ConvInputNHWCFilterHWCFOp,
-            linalg::DepthwiseConvInputNHWCFilterHWCOp>(op)) {
-      count.erase(count.begin());
+    size_t numParrallelLoops = getNumOuterParallelLoops(op);
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:615]: Oh Nice! This was exactly what I was trying to do. Should work well with Hanhan's PR #5136

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 4]
```diff
@@ -849,6 +849,23 @@ static void decideFusableLinalgOps(FuncOp funcOp) {
                             builder.getI64ArrayAttr(fusionGroups));
       }
     }
+
+    // As a second step mark all the element-wise linalg ops not fused as roots
+    // so that they get tiled and distributed.
+    for (linalg::LinalgOp linalgOp : linalgOps) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:855]: So the only reason we have the `MakeDispatchWorkgroupsOp` pattern is for the fall back right? In other words in the `ConvertToGPU` pass, the mapping of iterations of `linalg` operation to global invocation ID is not used in the Linalg to tensors path. That would be great, cause then we can move all that loop distribution logic all the way in Flow.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Looks awesome! I am super excited about this landing! It address a lot of hidden issues by enabling this path.

Just a nit on comments for setting the launch config. Once it is addressed, it is good to go.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp (hunk 1 above)
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp]: Good catch, you're right I had the comment backward. The problem is that when vectorization is off the second level of tiling is skipped and we end up mapping one element to one thread. So the real problem is that if we picked something different in LaunchConfig the number of workgroup is wrong. I updated the comment, hopefully this is clear.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp (hunk 4 above)
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:855]: Correct, there is still few cases that don't tile, I can look at those at some point. Yes for all those cases the `ConverToGPU` global invocation ID is not reached. It would be good to do the loop distribution in flow indeed. 
The main case I have seen remaining is when the linalg affine map are not projected permutation.

**Comment by ThomasRaoux** [comment]
Thanks Mahesh, please take another look.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp [hunk 5]
```diff
@@ -312,6 +312,50 @@ static LogicalResult getConfigForCooperativeMatmul(
...
+  // fall back to convertToGPU which will map one element to one thread. To
+  // avoid a mismatch in the number of workgroup dispatched, we pick a tile size
+  // to have one element per thread.
+  if (options.enableVectorization) {
+    sizes.append({4 * subgroupSize, 2 * subgroupSize});
+  }
+  sizes.push_back(subgroupSize);
+  // Use the first tile size that can divide the shape. If the shape is not
+  // aligned on any of the tile sizes pick the smallest tile of one element per
+  // thread.
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:340]: Please add a TODO here. FOr the linalg on tensors path this could actually be anything other than 1. It can be even on the old path with a few changes, but why bother when its going away. For now, leave a TODO and come back to it when we deprecate the old path.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp [hunk 6]
```diff
@@ -708,6 +752,27 @@ Optional<LaunchConfig> initGPULaunchConfig(
 #undef DISPATCH
   }
 
+  if (!rootOperation) {
+    for (linalg::LinalgOp linalgOp : linalgOps) {
+      if (auto op = dyn_cast<linalg::GenericOp>(linalgOp.getOperation())) {
+        if (getNumOuterParallelLoops(linalgOp) == 0 ||
+            llvm::any_of(linalgOp.getIndexingMaps(), [](AffineMap &map) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:759]: It might just be getting mixed with the old paths. This should not be needed actually.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp [hunk 7]
```diff
@@ -708,6 +752,27 @@ Optional<LaunchConfig> initGPULaunchConfig(
 #undef DISPATCH
   }
 
+  if (!rootOperation) {
+    for (linalg::LinalgOp linalgOp : linalgOps) {
+      if (auto op = dyn_cast<linalg::GenericOp>(linalgOp.getOperation())) {
+        if (getNumOuterParallelLoops(linalgOp) == 0 ||
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:758]: When we drop the old path, this will have to go through the new path.

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp (hunk 6 above)
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:759]: I still need this check otherwise some linalg op with weird affine map crash during tiling. I can look at why this happen after if you think this is not expected.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp (hunk 7 above)
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:758]: we still need this even in the new path right?

**Comment by ThomasRaoux** [comment]
I'll merge this PR, if there are still concerns with your comments I can address it after.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp (hunk 7 above)
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:759]: Sorry, bad comment. I was saying that this is basically saying dont do anything with 0-rank operations. But those need to be handled as well (i.e. lowered to loops albeit no-op) at some point. I am guessing this isnt tested on the new path. 
One thing to do here is just test what happens if a `linalg.generic` is just  0-rank operations.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp (hunk 6 above)
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:760]: Yeah, this is not expected. I dont see why it would carsh (at least would be good to know the affine map. Is this because of `reverse`? If so worth triaging upstream.

**Comment by hanhanW** [comment]
This seems to break compiling mobilenet_v2 in iree-android-benchmark.

To repro:
```
iree-translate --iree-mlir-to-vm-bytecode-module --iree-hal-target-backends=vulkan-spirv ~/mobilenet_v2.mlir -o a.vmfb --iree-spirv-enable-vectorization --iree-vulkan-target-triple=valhall-g77-unknown-android10
```

The mobilenet_v2.mlir can be downloaded from https://storage.googleapis.com/iree-model-artifacts/mobilenet-v2.tar.gz

It fails in ConvertToSPIRVPass. The input IR of ConvertToSPIRVPass: https://gist.github.com/hanhanW/1f1c5cb70ff1486c04783ea052e2703d

The error message is `failed to legalize operation 'vector.broadcast'`

**Comment by antiagainst** [comment]
Oh, a bunch of size-1 vectors are generated. Those aren't supported by SPIR-V. (They should just be scalars.)

@hanhanW: This is still checking the existing path? I think we should switch the benchmark tests to go via the new path. 

Also it would be nice to check in a faked weights MobileNet v2. Discovering breakage via benchmarking tests is not that good. :) 
 
@ThomasRaoux: can we revert this for now and have the issues fixed before landing it?

**Comment by ThomasRaoux** [comment]
> Oh, a bunch of size-1 vectors are generated. Those aren't supported by SPIR-V. (They should just be scalars.)
> 
> @hanhanW: This is still checking the existing path? I think we should switch the benchmark tests to go via the new path.
> 
> Also it would be nice to check in a faked weights MobileNet v2. Discovering breakage via benchmarking tests is not that good. :)
> 
> @ThomasRaoux: can we revert this for now and have the issues fixed before landing it?

Yes, let me revert it and I'll fix it offline.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp (hunk 6 above)
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:760]: Mahesh, yes the problem is with `reverse`, I'll work on fixing this part in a second step.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp (hunk 7 above)
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:759]: Good point, I can add a test for that.
